### ISSUE TITLE
[FEATURE] VLAN enhancement upgrading

### ISSUE THREAD

**Issue by yaocw2020** [harvester/harvester#2734]
**Is your feature request related to a problem? Please describe.**

Because there are some conflicts between the old VLAN implementation and the VLAN enhancement, especially the harvester installer part, we have to make some efforts for the VLAN enhancement upgrading.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by TachunLin** [comment]
### Test plan 
* Based on https://github.com/harvester/harvester/pull/2852#issue-1391209890 

#### Test scenario 1 (No vlan)
* Before upgrade: No network enabled and vlan created 
* After upgrade: Check new network feature and UI 

#### Test scenario 2 (harvester-mgmt vlan1)
* Before upgrade: Enable network on `harvester-mgmt` and create vlan `1` 
* After upgrade: 
  - Check the network bridge `mgmt-br` exists
  - Check there is vlan `1` created on default cluster network
  - Check the yaml content of vlan`1`
  - Check new network feature and UI 

#### Test scenario 3 (enps0 vlan)
* Before upgrade: Enable network on other NIC (eg. enps1) and create a vlan with valid id
* After upgrade: 
  - Check the network bridge `mgmt-br` exists
  - Check there is vlan with id created on related cluster network
  - Check the yaml content of available vlan
  - Check new network feature and UI 

#### Test scenario 4 (bond mode use config)
* Before upgrade: Use network config to bond multiple NICs (ipex-example)
* After upgrade: 
  - Check the network bridge exists
  - Check there is vlan with id created on related cluster network
  - Check the yaml content of available vlan
  - Check new network feature and UI

**Comment by TachunLin** [comment]
### (PASS) Test plan2: harvester-mgmt vlan1

#### Before upgrade (v1.0.3)
1. Enable network on `harvester-mgmt` 
  ![image](https://user-images.githubusercontent.com/29251855/197128989-7f241bd5-68b0-4af9-9c9b-8e11d46e36b4.png)
2. Create vlan id `1` 
  ![image](https://user-images.githubusercontent.com/29251855/197128677-26c97837-deb0-4f0a-b2cb-61c273d480a8.png)
3. Create two VMs, one set to vlan 1 and another use harvester-mgmt
  ![image](https://user-images.githubusercontent.com/29251855/197131772-0feeab74-978d-41fa-84d5-68718e1ab3ff.png)


#### After upgrade to v1.1.0-rc3
- [x] Check the network bridge `mgmt-br` exists
  ```
  7: mgmt-bo: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1500 qdisc noqueue master mgmt-br state UP group default qlen 1000
      link/ether ec:f4:bb:e5:61:7c brd ff:ff:ff:ff:ff:ff
  8: mgmt-br: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
      link/ether ec:f4:bb:e5:61:7c brd ff:ff:ff:ff:ff:ff
      inet 172.16.2.247/24 brd 172.16.2.255 scope global mgmt-br
         valid_lft forever preferred_lft forever
      inet 172.16.2.251/32 scope global mgmt-br
         valid_lft forever preferred_lft forever
      inet6 fe80::eef4:bbff:fee5:617c/64 scope link 
         valid_lft forever preferred_lft forever
  
  ```
- [x] Check there is vlan `1` created on default cluster network
  ![image](https://user-images.githubusercontent.com/29251855/197149446-87a09f67-fe36-452f-bbb0-64191aabfd9c.png)
  ![image](https://user-images.githubusercontent.com/29251855/197149511-d8893d32-16cb-4df5-97e5-b8d21e56ef80.png)

- [x] Check the yaml content of vlan`1`
  ```
  apiVersion: k8s.cni.cncf.io/v1
  kind: NetworkAttachmentDefinition
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"k8s.cni.cncf.io/v1","kind":"NetworkAttachmentDefinition","metadata":{"annotations":{"network.harvesterhci.io/route":"{\"mode\":\"auto\",\"cidr\":\"172.16.2.0/24\",\"gateway\":\"172.16.2.1\",\"connectivity\":\"true\"}"},"creationTimestamp":"2022-10-21T06:33:00Z","finalizers":["wrangler.cattle.io/harvester-network-nad-controller","wrangler.cattle.io/harvester-network-manager-nad-controller"],"generation":1,"labels":{"network.harvesterhci.io/type":"L2VlanNetwork","network.harvesterhci.io/vlan-id":"1"},"name":"vlan1","namespace":"default","resourceVersion":"578745","uid":"0f2fddd3-d789-424c-a273-dddb36463b0d"},"spec":{"config":"{\"cniVersion\":\"0.3.1\",\"name\":\"vlan1\",\"type\":\"bridge\",\"bridge\":\"mgmt-br\",\"promiscMode\":true,\"vlan\":1,\"ipam\":{}}"}}
      network.harvesterhci.io/route: '{"mode":"auto","cidr":"172.16.2.0/24","gateway":"172.16.2.1","connectivity":"true"}'
    creationTimestamp: "2022-10-21T06:33:00Z"
    finalizers:
    - wrangler.cattle.io/harvester-network-nad-controller
    - wrangler.cattle.io/harvester-network-manager-nad-controller
    generation: 2
    labels:
      network.harvesterhci.io/type: L2VlanNetwork
      network.harvesterhci.io/vlan-id: "1"
    managedFields:
    - apiVersion: k8s.cni.cncf.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations: {}
          f:labels:
            .: {}
            f:network.harvesterhci.io/type: {}
        f:spec: {}
      manager: harvester
      operation: Update
      time: "2022-10-21T06:33:00Z"
    - apiVersion: k8s.cni.cncf.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            f:network.harvesterhci.io/route: {}
          f:finalizers:
            .: {}
            v:"wrangler.cattle.io/harvester-network-manager-nad-controller": {}
            v:"wrangler.cattle.io/harvester-network-nad-controller": {}
          f:labels:
            f:network.harvesterhci.io/vlan-id: {}
      manager: harvester-network-controller
      operation: Update
      time: "2022-10-21T06:33:07Z"
    - apiVersion: k8s.cni.cncf.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            f:kubectl.kubernetes.io/last-applied-configuration: {}
        f:spec:
          f:config: {}
      manager: kubectl-client-side-apply
      operation: Update
      time: "2022-10-21T07:40:45Z"
    name: vlan1
    namespace: default
    resourceVersion: "625924"
    uid: 0f2fddd3-d789-424c-a273-dddb36463b0d
  spec:
    config: '{"cniVersion":"0.3.1","name":"vlan1","type":"bridge","bridge":"mgmt-br","promiscMode":true,"vlan":1,"ipam":{}}'
  ```
- [x] Check new network feature and UI 
  ![image](https://user-images.githubusercontent.com/29251855/197150387-fef7cfcc-301e-41e8-9a4d-581579c63e97.png)
  ![image](https://user-images.githubusercontent.com/29251855/197150437-f2b2be22-5015-4918-a79d-a4a38d23f7aa.png)

**Comment by TachunLin** [comment]
### (PASS) Test plan3: enps0 NIC with valid vlan

### Status 

- First trial Failed, tracked in https://github.com/harvester/harvester/issues/3021
- By using the [workaround](https://github.com/harvester/harvester/issues/3021#issuecomment-1288747614) to continue the remaining upgrade steps

#### Before upgrade: Enable network on other NIC (eg. enps1) and create a vlan with valid id
1. Enable network on `enp129s0` 
  ![image](https://user-images.githubusercontent.com/29251855/197371510-7f3bec68-3348-4123-ab4f-6b72507db4fc.png)

2. Create vlan id `91` on 
  ![image](https://user-images.githubusercontent.com/29251855/197371569-f7572b00-a262-4422-9406-80f77462f7b0.png)
  ![image](https://user-images.githubusercontent.com/29251855/197371584-5251ba57-e237-4342-b450-6320fcd49bd2.png)

3. Create two VMs, one set to vlan 91 and another use harvester-mgmt
  ![image](https://user-images.githubusercontent.com/29251855/197371971-4dd4b1f6-ecc7-4511-b262-467061204239.png)


#### After upgrade: 
- [x] Check the network bridge `mgmt-br` exists
  ```
  7: mgmt-bo: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1500 qdisc noqueue master mgmt-br state UP group default qlen 1000
      link/ether ec:f4:bb:e8:95:44 brd ff:ff:ff:ff:ff:ff
  8: mgmt-br: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
      link/ether ec:f4:bb:e8:95:44 brd ff:ff:ff:ff:ff:ff
      inet 172.16.2.44/24 brd 172.16.2.255 scope global mgmt-br
         valid_lft forever preferred_lft forever
      inet 172.16.2.250/32 scope global mgmt-br
         valid_lft forever preferred_lft forever
      inet6 fe80::eef4:bbff:fee8:9544/64 scope link 
         valid_lft forever preferred_lft forever
  ```
- [x] Check there is vlan with id created on related cluster network
![image](https://user-images.githubusercontent.com/29251855/197544329-3ff8170d-782c-464c-8145-c7bf282a0276.png)
![image](https://user-images.githubusercontent.com/29251855/197544401-fe314b50-db40-4076-8351-9651204df41b.png)

- [x] Check the yaml content of available vlan `91`
  ```
  apiVersion: k8s.cni.cncf.io/v1
  kind: NetworkAttachmentDefinition
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"k8s.cni.cncf.io/v1","kind":"NetworkAttachmentDefinition","metadata":{"annotations":{"network.harvesterhci.io/route":"{\"mode\":\"auto\",\"cidr\":\"172.16.91.0/24\",\"gateway\":\"172.16.91.1\",\"connectivity\":\"true\"}"},"creationTimestamp":"2022-10-23T03:14:08Z","finalizers":["wrangler.cattle.io/harvester-network-nad-controller","wrangler.cattle.io/harvester-network-manager-nad-controller"],"generation":1,"labels":{"network.harvesterhci.io/type":"L2VlanNetwork","network.harvesterhci.io/vlan-id":"91"},"name":"vlan91","namespace":"default","resourceVersion":"1556880","uid":"0f7b9ec8-2431-4161-97e5-8be82c7b97bd"},"spec":{"config":"{\"cniVersion\":\"0.3.1\",\"name\":\"vlan91\",\"type\":\"bridge\",\"bridge\":\"vlan-br\",\"promiscMode\":true,\"vlan\":91,\"ipam\":{}}"}}
      network.harvesterhci.io/route: '{"mode":"auto","cidr":"172.16.91.0/24","gateway":"172.16.91.1","connectivity":"true"}'
    creationTimestamp: "2022-10-23T03:14:08Z"
    finalizers:
    - wrangler.cattle.io/harvester-network-nad-controller
    - wrangler.cattle.io/harvester-network-manager-nad-controller
    generation: 2
    labels:
      network.harvesterhci.io/type: L2VlanNetwork
      network.harvesterhci.io/vlan-id: "91"
    managedFields:
    - apiVersion: k8s.cni.cncf.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations: {}
          f:labels:
            .: {}
            f:network.harvesterhci.io/type: {}
        f:spec: {}
      manager: harvester
      operation: Update
      time: "2022-10-23T03:14:08Z"
    - apiVersion: k8s.cni.cncf.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            f:network.harvesterhci.io/route: {}
          f:finalizers:
            .: {}
            v:"wrangler.cattle.io/harvester-network-manager-nad-controller": {}
            v:"wrangler.cattle.io/harvester-network-nad-controller": {}
          f:labels:
            f:network.harvesterhci.io/vlan-id: {}
      manager: harvester-network-controller
      operation: Update
      time: "2022-10-23T03:14:14Z"
    - apiVersion: k8s.cni.cncf.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            f:kubectl.kubernetes.io/last-applied-configuration: {}
        f:spec:
          f:config: {}
      manager: kubectl-client-side-apply
      operation: Update
      time: "2022-10-23T04:13:50Z"
    name: vlan91
    namespace: default
    resourceVersion: "1599988"
    uid: 0f7b9ec8-2431-4161-97e5-8be82c7b97bd
  spec:
    config: '{"cniVersion":"0.3.1","name":"vlan91","type":"bridge","bridge":"vlan-br","promiscMode":true,"vlan":91,"ipam":{}}'
  ```

- [x] Check new network feature and UI, vlan1 on `mgmt`, vlan91 on `vlan`
  ![image](https://user-images.githubusercontent.com/29251855/197545377-4e973615-4fd2-40b1-ac88-c593eb6ff43d.png)
  ![image](https://user-images.githubusercontent.com/29251855/197545451-92a30c9e-7979-4616-b20f-6ef7043285a7.png)

**Comment by TachunLin** [comment]
### (PASS) Test plan4: Bond mode using Harvester config file 

###
* First trial failed, `pending` at upgrade system service 
![image](https://user-images.githubusercontent.com/29251855/197749187-7b9fa451-a898-4ec2-90d7-3b403c93aa2f.png)

#### Before upgrade: Use network config to bond multiple NICs (ipex-example) 
1. Enable network on `harvester-vlan` bond interface 
  ![image](https://user-images.githubusercontent.com/29251855/197727249-f749ca9b-44bb-4414-a9bb-5ae30a3e9ab3.png)

2. Check the harvester-vlan link device status 
  ```
  harvester-node-0:~ # ip -d l show dev harvester-vlan
  6: harvester-vlan: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1500 qdisc noqueue master harvester-br0 state UP mode DEFAULT group default qlen 1000
      link/ether 52:54:00:b1:92:49 brd ff:ff:ff:ff:ff:ff promiscuity 1 minmtu 68 maxmtu 65535 
      bond mode balance-tlb active_slave ens7 miimon 100 updelay 0 downdelay 0 peer_notify_delay 0 use_carrier 1 arp_interval 0 arp_validate none arp_all_targets any primary_reselect always fail_over_mac none xmit_hash_policy layer2 resend_igmp 1 num_grat_arp 1 all_slaves_active 0 min_links 0 lp_interval 1 packets_per_slave 1 lacp_rate slow ad_select stable tlb_dynamic_lb 1 
      bridge_slave state forwarding priority 32 cost 100 hairpin off guard off root_block off fastleave off learning on flood on port_id 0x8001 port_no 0x1 designated_port 32769 designated_cost 0 designated_bridge 8000.52:54:0:b1:92:49 designated_root 8000.52:54:0:b1:92:49 hold_timer    0.00 message_age_timer    0.00 forward_delay_timer    0.00 topology_change_ack 0 config_pending 0 proxy_arp off proxy_arp_wifi off mcast_router 1 mcast_fast_leave off mcast_flood on mcast_to_unicast off neigh_suppress off group_fwd_mask 0 group_fwd_mask_str 0x0 vlan_tunnel off isolated off addrgenmode eui64 numtxqueues 16 numrxqueues 16 gso_max_size 65536 gso_max_segs 65535
  ```

3. Create several vlan based on `harvester-vlan` interface
  ![image](https://user-images.githubusercontent.com/29251855/197756515-e222f74e-d87d-4722-8abb-4e3e30569bf7.png)

4. Create a VM with vlan `1` network
  ![image](https://user-images.githubusercontent.com/29251855/197757137-1118ebc6-ffb7-4a8c-a6bf-39c8d3b1cf77.png)

5. The harvester-vlan config before upgrade
  ```
  harvester-vlan:
        interfaces:
        - name: ens7
          hwaddr: ""
        - name: ens8
          hwaddr: ""
        method: none
        ip: ""
        subnetmask: ""
        gateway: ""
        defaultroute: false
        bondoptions: {}
        mtu: 0
  ```

#### Finish upgrade
![image](https://user-images.githubusercontent.com/29251855/197764328-28aa9ecc-cce1-4c16-8094-0c288ee9246e.png)



#### After upgrade: 
- [x] Check the network bridge exists 
- [x] Check the cluster networks
  ![image](https://user-images.githubusercontent.com/29251855/197764683-e1668711-7863-43f3-91ab-c909220cdf0d.png)
  ![image](https://user-images.githubusercontent.com/29251855/197764959-163cb3a7-a803-42b9-9561-b91e5c44208c.png)

- [x] Check there is vlan with id created on related cluster network 
  ![image](https://user-images.githubusercontent.com/29251855/197765035-9fcbd7c1-5b51-472f-9bf8-10c980ec2c8a.png)

- [x] Check the yaml content of available vlan 
  ```
  apiVersion: k8s.cni.cncf.io/v1
  kind: NetworkAttachmentDefinition
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"k8s.cni.cncf.io/v1","kind":"NetworkAttachmentDefinition","metadata":{"annotations":{"network.harvesterhci.io/route":"{\"mode\":\"auto\",\"cidr\":\"192.168.0.0/24\",\"gateway\":\"192.168.0.1\",\"connectivity\":\"true\"}"},"creationTimestamp":"2022-10-25T10:58:49Z","finalizers":["wrangler.cattle.io/harvester-network-nad-controller","wrangler.cattle.io/harvester-network-manager-nad-controller"],"generation":1,"labels":{"network.harvesterhci.io/type":"L2VlanNetwork","network.harvesterhci.io/vlan-id":"1"},"name":"vlan1","namespace":"default","resourceVersion":"11015","uid":"6cadbe68-7c92-4be3-864b-ab2e67674590"},"spec":{"config":"{\"cniVersion\":\"0.3.1\",\"name\":\"vlan1\",\"type\":\"bridge\",\"bridge\":\"vlan-br\",\"promiscMode\":true,\"vlan\":1,\"ipam\":{}}"}}
      network.harvesterhci.io/route: '{"mode":"auto","cidr":"192.168.0.0/24","gateway":"192.168.0.1","connectivity":"true"}'
    creationTimestamp: "2022-10-25T10:58:49Z"
    finalizers:
    - wrangler.cattle.io/harvester-network-nad-controller
    - wrangler.cattle.io/harvester-network-manager-nad-controller
    generation: 2
    labels:
      network.harvesterhci.io/type: L2VlanNetwork
      network.harvesterhci.io/vlan-id: "1"
    managedFields:
    - apiVersion: k8s.cni.cncf.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations: {}
          f:labels:
            .: {}
            f:network.harvesterhci.io/type: {}
        f:spec: {}
      manager: harvester
      operation: Update
      time: "2022-10-25T10:58:49Z"
    - apiVersion: k8s.cni.cncf.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            f:network.harvesterhci.io/route: {}
          f:finalizers:
            .: {}
            v:"wrangler.cattle.io/harvester-network-manager-nad-controller": {}
            v:"wrangler.cattle.io/harvester-network-nad-controller": {}
          f:labels:
            f:network.harvesterhci.io/vlan-id: {}
      manager: harvester-network-controller
      operation: Update
      time: "2022-10-25T10:58:55Z"
    - apiVersion: k8s.cni.cncf.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            f:kubectl.kubernetes.io/last-applied-configuration: {}
        f:spec:
          f:config: {}
      manager: kubectl-client-side-apply
      operation: Update
      time: "2022-10-25T11:20:55Z"
    name: vlan1
    namespace: default
    resourceVersion: "27069"
    uid: 6cadbe68-7c92-4be3-864b-ab2e67674590
  spec:
    config: '{"cniVersion":"0.3.1","name":"vlan1","type":"bridge","bridge":"vlan-br","promiscMode":true,"vlan":1,"ipam":{}}'
  
  ```

- [x] Check the harvester network config vlan content 
  ```
  harvester-vlan:
        interfaces:
        - name: ens7
          hwaddr: ""
        - name: ens8
          hwaddr: ""
        method: none
        ip: ""
        subnetmask: ""
        gateway: ""
        defaultroute: false
        bondoptions: {}
        mtu: 0
  ```

- [x] Check new network feature and UI

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.1.0-rc3`. Close this issue.

### Result
Verified the VLAN upgrade enhancement can PASS the following test plans:
1. Test plan2: harvester-mgmt vlan1 
2. Test plan3: enps0 NIC with valid vlan
3. Test plan4: Bond mode using Harvester config file

### Test Information
* Test Environment: 1 ~ 3 nodes Harvester cluster on bare machine and local kvm
* Harvester version: `v1.1.0-rc3`

### Verify Steps
https://github.com/harvester/harvester/issues/2734#issuecomment-1286638172
https://github.com/harvester/harvester/issues/2734#issuecomment-1286640032
https://github.com/harvester/harvester/issues/2734#issuecomment-1289100912
